File: tests/Unit/PointwiseFunctions/AnalyticSolutions/ForceFree/AlfvenWave.py
# Distributed under the MIT License.
# See LICENSE.txt for details.

import numpy as np


def solution_wave_frame(coords_prime):
    electric_field = coords_prime * 0.0
    magnetic_field = coords_prime * 0.0

    x = coords_prime[0]

    magnetic_field[0] = 1.0
    magnetic_field[1] = 1.0
    magnetic_field[2] = np.where(
        x > -0.1, 1.0 + 0.15 * (1.0 + np.sin(5 * np.pi * x)), 1.0
    )
    magnetic_field[2] = np.where(x > 0.1, 1.3, magnetic_field[2])

    electric_field[0] = -magnetic_field[2]
    electric_field[2] = 1.0

    return (electric_field, magnetic_field)


def lorentz_transform_em_field(v, em_fields_tuple):
    # transforms electric and magnetic field to their values in the frame
    # moving with the speed (v, 0, 0) with respect to the original frame
    electric_field, magnetic_field = em_fields_tuple

    electric_field_prime = electric_field * 1.0
    magnetic_field_prime = magnetic_field * 1.0

    lorentz_factor = 1.0 / np.sqrt(1 - v**2)

    electric_field_prime[1] = lorentz_factor * (
        electric_field[1] - v * magnetic_field[2]
    )
    electric_field_prime[2] = lorentz_factor * (
        electric_field[2] + v * magnetic_field[1]
    )

    magnetic_field_prime[1] = lorentz_factor * (
        magnetic_field[1] + v * electric_field[2]
    )
    magnetic_field_prime[2] = lorentz_factor * (
        magnetic_field[2] - v * electric_field[1]
    )

    return (electric_field_prime, magnetic_field_prime)


def TildeE(x, t, wave_speed):
    lorentz_factor = 1.0 / np.sqrt(1 - wave_speed**2)
    x0 = x * 1.0
    x0[0] -= wave_speed * t

    return lorentz_transform_em_field(
        -wave_speed, solution_wave_frame(lorentz_factor * x0)
    )[0]


def TildeB(x, t, wave_speed):
    lorentz_factor = 1.0 / np.sqrt(1 - wave_speed**2)
    x0 = x * 1.0
    x0[0] -= wave_speed * t
    return lorentz_transform_em_field(
        -wave_speed, solution_wave_frame(lorentz_factor * x0)
    )[1]


def TildePsi(x, t, wave_speed):
    return x[0] * 0


def TildePhi(x, t, wave_speed):
    return x[0] * 0


def solution_charge_wave_frame(x):
    q = np.where(x < -0.1, 0.0, -0.75 * np.pi * np.cos(5.0 * np.pi * x))
    q = np.where(x > 0.1, 0.0, q)
    return q


def TildeQ(x, t, wave_speed):
    lorentz_factor = 1.0 / np.sqrt(1 - wave_speed**2)
    return lorentz_factor * solution_charge_wave_frame(
        lorentz_factor * (x[0] - wave_speed * t)
    )

File: src/PointwiseFunctions/AnalyticSolutions/ForceFree/AlfvenWave.cpp
// Distributed under the MIT License.
// See LICENSE.txt for details.

#include "PointwiseFunctions/AnalyticSolutions/ForceFree/AlfvenWave.hpp"

#include <cstddef>
#include <memory>
#include <pup.h>

#include "DataStructures/DataVector.hpp"
#include "DataStructures/Tensor/Tensor.hpp"
#include "Options/ParseError.hpp"
#include "PointwiseFunctions/InitialDataUtilities/InitialData.hpp"
#include "Utilities/ConstantExpressions.hpp"
#include "Utilities/ErrorHandling/Assert.hpp"
#include "Utilities/MakeWithValue.hpp"
#include "Utilities/TaggedTuple.hpp"

namespace ForceFree::Solutions {

AlfvenWave::AlfvenWave(const double wave_speed, const Options::Context& context)
    : wave_speed_(wave_speed) {
  if (abs(wave_speed_) >= 1.0) {
    PARSE_ERROR(context,
                "The wave speed ("
                    << wave_speed_
                    << ") must be bigger than -1.0 and smaller than 1.0");
  }
  lorentz_factor_ = 1.0 / sqrt(1.0 - square(wave_speed_));
}

AlfvenWave::AlfvenWave(CkMigrateMessage* msg) : InitialData(msg) {}

std::unique_ptr<evolution::initial_data::InitialData> AlfvenWave::get_clone()
    const {
  return std::make_unique<AlfvenWave>(*this);
}

void AlfvenWave::pup(PUP::er& p) {
  InitialData::pup(p);
  p | wave_speed_;
  p | lorentz_factor_;
  p | background_spacetime_;
}

PUP::able::PUP_ID AlfvenWave::my_PUP_ID = 0;

DataVector AlfvenWave::wave_profile(const DataVector& x_prime) {
  // Compute the B_z'(=-E_x') at the rest frame of the wave
  auto result = make_with_value<DataVector>(x_prime, 1.0);
  for (size_t i = 0; i < result.size(); ++i) {
    const double& xi = x_prime[i];
    if (xi > -0.1) {
      if (xi <= 0.1) {
        result[i] = 1.15 + 0.15 * sin(5 * M_PI * xi);
      } else {
        result[i] = 1.3;
      }
    }
  }
  return result;
}

DataVector AlfvenWave::charge_density(const DataVector& x_prime) {
  // Compute the charge density q = Div(E) at the rest frame of the wave
  auto result = make_with_value<DataVector>(x_prime, 0.0);
  for (size_t i = 0; i < result.size(); ++i) {
    const double& xi = x_prime[i];
    if ((xi > -0.1) and (xi <= 0.1)) {
      result[i] = -0.75 * M_PI * cos(5.0 * M_PI * xi);
    }
  }
  return result;
}

tuples::TaggedTuple<Tags::TildeE> AlfvenWave::variables(
    const tnsr::I<DataVector, 3>& x, double t,
    tmpl::list<Tags::TildeE> /*meta*/) const {
  const auto& x_coord = get<0>(x);
  auto electric_field = make_with_value<tnsr::I<DataVector, 3>>(x, 0.0);

  // E_x
  get<0>(electric_field) =
      -wave_profile(lorentz_factor_ * (x_coord - wave_speed_ * t));
  // E_y
  get<1>(electric_field) =
      -lorentz_factor_ * wave_speed_ * get<0>(electric_field);
  // E_z
  get<2>(electric_field) = lorentz_factor_ * (1.0 - wave_speed_);

  return electric_field;
}

tuples::TaggedTuple<Tags::TildeB> AlfvenWave::variables(
    const tnsr::I<DataVector, 3>& x, double t,
    tmpl::list<Tags::TildeB> /*meta*/) const {
  const auto& x_coord = get<0>(x);
  auto magnetic_field = make_with_value<tnsr::I<DataVector, 3>>(x, 1.0);
  // note) B_x = 1.0
  // B_y
  get<1>(magnetic_field) = lorentz_factor_ * (1.0 - wave_speed_);
  // B_z
  get<2>(magnetic_field) =
      lorentz_factor_ *
      wave_profile(lorentz_factor_ * (x_coord - wave_speed_ * t));

  return magnetic_field;
}

tuples::TaggedTuple<Tags::TildePsi> AlfvenWave::variables(
    const tnsr::I<DataVector, 3>& x, double /*t*/,
    tmpl::list<Tags::TildePsi> /*meta*/) {
  return {make_with_value<Scalar<DataVector>>(x, 0.0)};
}

tuples::TaggedTuple<Tags::TildePhi> AlfvenWave::variables(
    const tnsr::I<DataVector, 3>& x, double /*t*/,
    tmpl::list<Tags::TildePhi> /*meta*/) {
  return {make_with_value<Scalar<DataVector>>(x, 0.0)};
}

tuples::TaggedTuple<Tags::TildeQ> AlfvenWave::variables(
    const tnsr::I<DataVector, 3>& x, double t,
    tmpl::list<Tags::TildeQ> /*meta*/) const {
  return {Scalar<DataVector>{
      lorentz_factor_ *
      charge_density(lorentz_factor_ * (get<0>(x) - wave_speed_ * t))}};
}

bool operator==(const AlfvenWave& lhs, const AlfvenWave& rhs) {
  return lhs.wave_speed_ == rhs.wave_speed_ and
         lhs.lorentz_factor_ == rhs.lorentz_factor_ and
         lhs.background_spacetime_ == rhs.background_spacetime_;
}

bool operator!=(const AlfvenWave& lhs, const AlfvenWave& rhs) {
  return not(lhs == rhs);
}

}  // namespace ForceFree::Solutions
